src/main/java/edu/byu/cs/controller/SubmissionController.java

```diff
@@ -558,29 +566,66 @@ public static void approveSubmission(
             LOGGER.warn("Approving submissions did not affect any submissions. Something probably went wrong.");
         }
 
-        // Determine approvedScore
-        if (approvedScore == null) {
-            float bestScoreForPhase = submissionDao.getBestScoreForPhase(studentNetId, phase);
-            if (bestScoreForPhase < 0.0f) {
-                throw new RuntimeException("Cannot determine best score for phase without any submissions in the phase.");
+        if (submissionToUse == null) {
+            submissionToUse = submissionDao.getBestSubmissionForPhase(studentNetId, phase);
+            if (submissionToUse == null) {
+                throw new RuntimeException("No submission was provided nor found for phase " + phase + " with user " + studentNetId);
             }
-            approvedScore = SubmissionHelper.prepareModifiedScore(bestScoreForPhase, scoreVerification);
         }
-        if (approvedScore <= 0.0f) {
-            throw new RuntimeException("Cannot set grade without a positive approvedScore!");
+        if (approvedScore == null) {
+            approvedScore = SubmissionHelper.prepareModifiedScore(submissionToUse.score(), penaltyPct);
```

Comment:
Somewhere in here we are double counting penalties. The original score already appears to be reduced.

Reply:
Thanks for observing that. I'll double check that.

Reply:
There were a couple issues previous code that have all been cleaned up. Examples of known issues (now fixed) include:
* Grabbing the worst submission instead of the best submission
* Using the first submission as a baseline instead of the best submission

Besides merely fixing these issues, the code in nearby areas has been completely restructured to reduce the likelihood of similar issues occurring again. This refactoring includes the following adjustments:
1. Splitting up long methods into smaller, focused methods
2. Adjusting method call signatures to provide defaults are require the minimum required data
3. Removing duplicated code and relying on existing methods
4. Adding JavaDoc comments to clearly document the purposes of each method in the transformation process